src/lib/comms/sol-http-common.c

```diff
@@ -123,3 +120,453 @@ sol_http_escape_string(char **escaped, const char *value)
+    const struct sol_http_param *params)
+{
+    struct sol_buffer buf;
+    struct sol_http_param_value *iter;
+    char *encoded_key, *encoded_value;
+    uint16_t idx;
+    bool first = true;
+    int r;
+
+    SOL_NULL_CHECK(encoded_params, -EINVAL);
+    SOL_NULL_CHECK(params, -EINVAL);
+
+    if (type != SOL_HTTP_PARAM_QUERY_PARAM &&
+        type != SOL_HTTP_PARAM_POST_FIELD &&
+        type != SOL_HTTP_PARAM_COOKIE &&
+        type != SOL_HTTP_PARAM_HEADER) {
+        return -EINVAL;
+    }
+
+    sol_buffer_init(&buf);
+
+    encoded_key = encoded_value = NULL;
+    SOL_HTTP_PARAM_FOREACH_IDX (params, iter, idx) {
+        if (iter->type != type)
+            continue;
+
+        r = sol_http_escape_string(&encoded_key, iter->value.key_value.key);
+        SOL_INT_CHECK_GOTO(r, < 0, clean_up);
+
+        r = sol_http_escape_string(&encoded_value, iter->value.key_value.value);
+        SOL_INT_CHECK_GOTO(r, < 0, clean_up);
+
+        if (type == SOL_HTTP_PARAM_COOKIE) {
+            r = sol_buffer_append_printf(&buf, "%s%s=%s;", first ? "" : " ",
+                encoded_key, encoded_value);
+        } else {
+            r = sol_buffer_append_printf(&buf, "%s%s=%s", first ? "" : "&",
+                encoded_key, encoded_value);
+        }
+
+        SOL_INT_CHECK_GOTO(r, < 0, clean_up);
+
+        first = false;
+        free(encoded_key);
+        free(encoded_value);
+        encoded_value = encoded_key = NULL;
+    }
+
+    r = sol_buffer_ensure_nul_byte(&buf);
+    SOL_INT_CHECK_GOTO(r, < 0, err_exit);
+    *encoded_params = sol_buffer_steal(&buf, NULL);
+
+    sol_buffer_fini(&buf);
+
+    return 0;
+
+clean_up:
+    free(encoded_key);
+    free(encoded_value);
+err_exit:
+    sol_buffer_fini(&buf);
+    return r;
+}
+
+SOL_API int
+sol_http_create_uri(char **uri, const char *protocol,
+    const char *server, const char *path, const char *fragment, int port,
+    const struct sol_http_param *params)
+{
+    struct sol_buffer buf;
+    char *encoded_params = NULL;
+    int r;
+
+    SOL_NULL_CHECK(uri, -EINVAL);
+    SOL_NULL_CHECK(server, -EINVAL);
+
+    if (params) {
+        r = sol_http_encode_params(&encoded_params, SOL_HTTP_PARAM_QUERY_PARAM,
+            params);
+        SOL_INT_CHECK(r, < 0, r);
+    }
+
+    sol_buffer_init(&buf);
+
+    if (protocol)
+        r = sol_buffer_append_printf(&buf, "%s://%s", protocol, server);
+    else
+        r = sol_buffer_append_printf(&buf, "http://%s", server);
+    SOL_INT_CHECK_GOTO(r, < 0, exit);
+
+    if (port != 80) {
+        r = sol_buffer_append_printf(&buf, ":%d", port);
+        SOL_INT_CHECK_GOTO(r, < 0, exit);
+    }
+
+    if (path && (strlen(path) >= 1 && path[0] != '/')) {
+        r = sol_buffer_append_printf(&buf, "%s%s",
+            path[0] != '/' ? "/" : "", path);
+        SOL_INT_CHECK_GOTO(r, < 0, exit);
+    }
+
+    if (encoded_params && *encoded_params) {
+        r = sol_buffer_append_printf(&buf, "?%s", encoded_params);
```

Comment:
if you change `sol_http_encode_params()` to take an already-initialized buffer where you just `sol_buffer_append()`then you will simplify it here, no need to `encode_params` as a separate memory that is later printf to the uri buffer. You do all in a single buffer.

Just be sure to document such behavior in the `sol-http.h` file.